torch/csrc/api/include/torch/optimizers.h

```diff
@@ -13,6 +13,7 @@ class OptimizerImpl {
   virtual ~OptimizerImpl() = default;
   virtual void init_state() {}
   virtual void step() = 0;
+  virtual at::Scalar step(std::function<at::Scalar()> closure) = 0;
```

Comment:
I think I prefer just one function, but where the signature looks like:

```c++
virtual at::Scalar step(std::function<at::Scalar()> closure = OptimizerImpl::NoLoss) = 0;
```

and define NoLoss to just return an empty scalar. Keep in mind you'll have to define the default again in each of the subclasses as well. It's slightly cleaner than having a separate step() function that users might have to remember to implement.